File: Backend/services/commerce/financial-services-java/ledger-service/src/main/java/com/baalvion/ledger/domain/JournalEntry.java
package com.baalvion.ledger.domain;

import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

/**
 * JournalEntry: Immutable double-entry bookkeeping record
 *
 * Every transaction posts a pair: debit one account, credit another.
 * Ensures ledger always balances: sum(debits) = sum(credits).
 */
@Entity
@Table(
  name = "journal_entries",
  schema = "ledger",
  indexes = {
    @Index(name = "idx_tenant_status_date", columnList = "tenant_id,status,posted_at DESC"),
    @Index(name = "idx_debit_account", columnList = "debit_account_id,tenant_id"),
    @Index(name = "idx_credit_account", columnList = "credit_account_id,tenant_id"),
    @Index(name = "idx_tenant_date", columnList = "tenant_id,posted_at DESC")
  },
  uniqueConstraints = {
    @UniqueConstraint(name = "uk_transaction_ref", columnNames = {"transaction_ref", "tenant_id"})
  }
)
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class JournalEntry {

  @Id
  @Column(columnDefinition = "uuid")
  private UUID id;

  @NotNull(message = "Tenant ID required")
  @Column(columnDefinition = "uuid", nullable = false)
  private UUID tenantId;

  @NotBlank(message = "Transaction reference required")
  @Size(max = 64, message = "Transaction reference must not exceed 64 characters")
  @Column(length = 64, nullable = false)
  private String transactionRef;

  @NotNull(message = "Debit account required")
  @Column(columnDefinition = "uuid", nullable = false)
  private UUID debitAccountId;

  @NotNull(message = "Credit account required")
  @Column(columnDefinition = "uuid", nullable = false)
  private UUID creditAccountId;

  @NotNull(message = "Amount required")
  @DecimalMin(value = "0.01", message = "Amount must be greater than zero")
  @Digits(integer = 15, fraction = 4, message = "Amount must have max 15 digits and 4 decimals")
  @Column(precision = 19, scale = 4, nullable = false)
  private BigDecimal amount;

  @NotBlank(message = "Currency required")
  @Size(min = 3, max = 3, message = "Currency must be ISO 4217 code (3 chars)")
  @Column(length = 3, nullable = false)
  private String currency;

  @NotNull(message = "Entry type required")
  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  private EntryType entryType;

  @NotNull(message = "Status required")
  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  private EntryStatus status;

  @Column(columnDefinition = "TEXT")
  private String description;

  @Column(columnDefinition = "uuid")
  private UUID relatedTransactionId;

  @Column(nullable = true)
  private LocalDateTime postedAt;

  @Column(nullable = true)
  private LocalDateTime reversedAt;

  @Column(columnDefinition = "jsonb", nullable = false)
  private String metadata;

  @CreationTimestamp
  @Column(nullable = false, updatable = false)
  private LocalDateTime createdAt;

  @UpdateTimestamp
  @Column(nullable = false)
  private LocalDateTime updatedAt;

  @Version
  private Long version;

  public enum EntryType {
    PAYMENT,
    FEE,
    REVERSAL,
    SETTLEMENT,
    ESCROW,
    REFUND,
    ADJUSTMENT
  }

  public enum EntryStatus {
    PENDING,
    POSTED,
    REVERSED
  }

  @PrePersist
  protected void onCreate() {
    if (id == null) {
      id = UUID.randomUUID();
    }
    if (metadata == null) {
      metadata = "{}";
    }
  }
}

File: Backend/services/commerce/ledger-service/models/JournalEntry.js
const { DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  const JournalEntry = sequelize.define(
    'JournalEntry',
    {
      id: {
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV4,
        primaryKey: true,
      },
      tenantId: {
        type: DataTypes.UUID,
        allowNull: false,
        index: true,
      },
      transactionRef: {
        type: DataTypes.STRING(64),
        allowNull: false,
        comment: 'Business reference (idempotency key)',
      },
      debitAccountId: {
        type: DataTypes.UUID,
        allowNull: false,
        comment: 'Account debited',
      },
      creditAccountId: {
        type: DataTypes.UUID,
        allowNull: false,
        comment: 'Account credited',
      },
      amount: {
        type: DataTypes.DECIMAL(19, 4),
        allowNull: false,
        comment: 'Amount in minor units (e.g., cents)',
      },
      currency: {
        type: DataTypes.CHAR(3),
        allowNull: false,
        defaultValue: 'USD',
        comment: 'ISO 4217 currency code',
      },
      entryType: {
        type: DataTypes.ENUM(
          'PAYMENT',
          'FEE',
          'REVERSAL',
          'SETTLEMENT',
          'ESCROW',
          'REFUND',
          'ADJUSTMENT'
        ),
        allowNull: false,
      },
      status: {
        type: DataTypes.ENUM('PENDING', 'POSTED', 'REVERSED'),
        defaultValue: 'PENDING',
      },
      description: {
        type: DataTypes.TEXT,
        allowNull: true,
      },
      relatedTransactionId: {
        type: DataTypes.UUID,
        allowNull: true,
        comment: 'Link to originating transaction',
      },
      postedAt: {
        type: DataTypes.DATE,
        allowNull: true,
        comment: 'Immutable posting timestamp',
      },
      reversedAt: {
        type: DataTypes.DATE,
        allowNull: true,
      },
      metadata: {
        type: DataTypes.JSONB,
        defaultValue: {},
        comment: 'Additional context (trade ID, RFQ ID, etc.)',
      },
      createdAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
      },
      updatedAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
      },
    },
    {
      sequelize,
      modelName: 'JournalEntry',
      tableName: 'journal_entries',
      schema: 'ledger',
      timestamps: true,
      indexes: [
        {
          fields: ['tenantId', 'status', 'postedAt'],
          name: 'idx_journal_entries_tenant_status_date',
        },
        {
          fields: ['transactionRef', 'tenantId'],
          unique: true,
          name: 'idx_journal_entries_unique_ref_per_tenant',
        },
        {
          fields: ['debitAccountId', 'tenantId'],
          name: 'idx_journal_entries_debit_account',
        },
        {
          fields: ['creditAccountId', 'tenantId'],
          name: 'idx_journal_entries_credit_account',
        },
      ],
    }
  );

  return JournalEntry;
};
